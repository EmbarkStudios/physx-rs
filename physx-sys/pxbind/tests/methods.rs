struct Functions {
    cpp: String,
    rust: String,
}

fn gen_functions(which: &str, to_emit: &'static [&str]) -> anyhow::Result<Functions> {
    let ast = pxbind::get_parsed_ast(format!("tests/data/functions/{which}"))?;

    let mut consumer = pxbind::consumer::AstConsumer::default();
    consumer.consume(&ast)?;

    let record_filter = |_rb: &pxbind::consumer::RecBinding<'_>| false;
    let enum_filter = |_eb: &pxbind::consumer::EnumBinding<'_>| false;
    let func_filter = |fb: &pxbind::consumer::FuncBinding<'_>| {
        if let Some(cname) = fb.owning_class() {
            to_emit.is_empty() || to_emit.iter().any(|te| *te == cname)
        } else {
            false
        }
    };

    let generator = pxbind::generator::Generator {
        record_filter: Box::new(record_filter),
        enum_filter: Box::new(enum_filter),
        func_filter: Box::new(func_filter),
        ..Default::default()
    };

    let cpp = {
        let mut buffer = Vec::new();
        generator.generate_cpp_functions(&consumer, &mut buffer, 0)?;
        String::from_utf8(buffer)?
    };

    let rust = {
        let mut buffer = Vec::new();
        generator.generate_rust_functions(&consumer, &mut buffer, 0)?;
        String::from_utf8(buffer)?
    };

    Ok(Functions { cpp, rust })
}

/// Verifies that reference fields are reflected in the size and layout, but aren't
/// actually publicly exposed
#[test]
fn methods_complex() {
    let mo = gen_functions("methods_complex.h", &["PxShape", "PxThreadImpl"]).unwrap();

    insta::assert_snapshot!(mo.cpp);
    insta::assert_snapshot!(mo.rust);
}

/// Verifies constructors of various kinds are properly bound, and properly
/// named to avoid clashes
#[test]
fn constructors() {
    let mo = gen_functions("constructors.h", &["PxBoxControllerDesc", "PxFilterData"]).unwrap();

    insta::assert_snapshot!(mo.cpp);
    insta::assert_snapshot!(mo.rust);
}

/// Verifies destructors are properly emitted or omitted
#[test]
fn destructor() {
    let mo = gen_functions(
        "destructors.h",
        &[
            "PxActor",
            "PxJointLimitParameters",
            "PxJointAngularLimitPair",
        ],
    )
    .unwrap();

    insta::assert_snapshot!(mo.cpp);
    insta::assert_snapshot!(mo.rust);
}
