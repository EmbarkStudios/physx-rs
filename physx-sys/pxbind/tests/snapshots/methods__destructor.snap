---
source: physx-sys/pxbind/tests/methods.rs
expression: mo.cpp
---
extern "C" {
    physx_PxJointLimitParameters_Pod* PxJointLimitParameters_new_alloc() {
        auto return_val = new PxJointLimitParameters();
        auto return_val_pod = reinterpret_cast<physx_PxJointLimitParameters_Pod*>(return_val);
        return return_val_pod;
    }

    bool PxJointLimitParameters_isValid(physx_PxJointLimitParameters_Pod const* self__pod) {
        auto self_ = reinterpret_cast<physx::PxJointLimitParameters const*>(self__pod);
        bool return_val = self_->isValid();
        return return_val;
    }

    bool PxJointLimitParameters_isSoft(physx_PxJointLimitParameters_Pod const* self__pod) {
        auto self_ = reinterpret_cast<physx::PxJointLimitParameters const*>(self__pod);
        bool return_val = self_->isSoft();
        return return_val;
    }

    physx_PxJointAngularLimitPair_Pod PxJointAngularLimitPair_new(float lowerLimit, float upperLimit, float contactDist_deprecated) {
        PxJointAngularLimitPair return_val(lowerLimit, upperLimit, contactDist_deprecated);
        physx_PxJointAngularLimitPair_Pod return_val_pod;
        memcpy(&return_val_pod, &return_val, sizeof(return_val_pod));
        return return_val_pod;
    }

    bool PxJointAngularLimitPair_isValid(physx_PxJointAngularLimitPair_Pod const* self__pod) {
        auto self_ = reinterpret_cast<physx::PxJointAngularLimitPair const*>(self__pod);
        bool return_val = self_->isValid();
        return return_val;
    }

    void PxJointAngularLimitPair_delete(physx_PxJointAngularLimitPair_Pod* self__pod) {
        auto self_ = reinterpret_cast<physx::PxJointAngularLimitPair*>(self__pod);
        delete self_;
    }

}

